servicetalk-concurrent-api/src/testFixtures/java/io/servicetalk/concurrent/api/TestCollectingPublisherSubscriber.java

```diff
@@ -0,0 +1,213 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import io.servicetalk.concurrent.PublisherSource.Subscriber;
+import io.servicetalk.concurrent.PublisherSource.Subscription;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.internal.ThrowableUtils.unknownStackTrace;
+
+/**
+ * A {@link Subscriber} that enqueues {@link #onNext(Object)} and terminal signals while providing blocking methods
+ * to consume these events. There are two approaches to using this class:
+ * <pre>
+ *     TestCollectingPublisherSubscriber&lt;String&gt; sub = new TestCollectingPublisherSubscriber&lt;&gt;();
+ *
+ *     // Approach 1 - verify individual items sequentially.
+ *     String s = sub.takeOnNext();
+ *     // verify s
+ *     sub.awaitOnComplete(); // this will verify that all onNext signals have been consumed
+ *
+ *     // Approach 2 - wait for terminal, verify items in bulk.
+ *     sub.awaitOnComplete(false); // wait for the terminal signal, ignore if there are unconsumed onNext signals.
+ *     List&lt;String&gt; onNextSignals = sub.pollAllOnNext();
+ *     // verify all onNextSignals occurred in the expected order
+ * </pre>
+ * @param <T> The type of data in {@link #onNext(Object)}.
+ */
+public final class TestCollectingPublisherSubscriber<T> implements Subscriber<T> {
+    private static final Object NULL_ON_NEXT = new Object();
+    private static final Throwable ON_COMPLETE = unknownStackTrace(new RuntimeException("onComplete"),
+            TestCollectingPublisherSubscriber.class, "onComplete");
+    private final BlockingQueue<Object> items = new LinkedBlockingQueue<>();
+    private final CountDownLatch onTerminalLatch = new CountDownLatch(1);
+    private final CountDownLatch onSubscribeLatch = new CountDownLatch(1);
+    @Nullable
+    private Throwable onTerminal;
+    @Nullable
+    private Subscription subscription;
+
+    @Override
+    public void onSubscribe(final Subscription subscription) {
+        this.subscription = subscription;
+        onSubscribeLatch.countDown();
+    }
+
+    @Override
+    public void onNext(@Nullable final T t) {
+        items.add(t == null ? NULL_ON_NEXT : t);
```

Comment:
Consider adding 

`assert onTerminal == null;`

to make sure there is no `onNext` after a terminal since we verify consumed items post terminal with
 `awaitOn[Error/Complete](boolean verifyOnNextConsumed)`